### PR TITLE
Refactor `needs-parens.js`

### PR DESCRIPTION
**Pull request by fisker**
Similar to @brodybits did in #8878, remove unnecessary conditions, move name check to the first, to avoid access node properties.

- [ ] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/language-js/needs-parens.js [hunk 1]
```diff
@@ -531,7 +531,13 @@ function needsParens(path, options) {
         (grandParent.init === parent || grandParent.update === parent)
       ) {
         return false;
-      } else if (name === "value" && parent.type === "Property") {
+      } else if (
+        name === "value" &&
+        parent.type === "Property" &&
+        grandParent &&
+        grandParent.type === "ObjectPattern" &&
+        grandParent.properties.includes(parent)
```
Comment (thorn0) [inline src/language-js/needs-parens.js:539]: BTW, `ObjectPattern` doesn't have any other child nodes, only `properties`, so this `includes` seems unneeded.

Comment (fisker) [reply src/language-js/needs-parens.js:539]: I know that, just for safety.

Comment (fisker) [reply src/language-js/needs-parens.js:539]: By the way, do you think adopt `esquery` is a good idea?

So we can do something like

```js
isNodeMatch(node, 'ObjectPattern > Property.properties > AssignmentExpression.value');
```
I'm a little worry about performance

Comment (thorn0) [reply src/language-js/needs-parens.js:539]: Looks interesting, but I have doubts about the performance too.

Comment (fisker) [reply src/language-js/needs-parens.js:539]: Another off topic one, I have been thinking about use esquery selector in `embeddedLanguageFormatting` option few weeks ago, :smile:

**Review by thorn0 — APPROVED** [review]

**Comment by fisker** [comment]
Wrote a eslint rule for it, and found more :smile:

**Review by evilebottnawi — APPROVED** [review]

File: src/language-js/needs-parens.js [hunk 2]
```diff
@@ -488,35 +495,32 @@ function needsParens(path, options) {
...
+        name === "key" &&
         parent.type === "ClassProperty" &&
-        parent.key === node &&
         parent.computed
       ) {
         return false;
-      } else if (
-        parent.type === "TSPropertySignature" &&
-        parent.name === node
-      ) {
```
Comment (thorn0) [inline src/language-js/needs-parens.js:510]: This seems to be an artifact left from older versions of `typescript-eslint`. `TSPropertySignature` doesn't have the `name` property anymore. See https://github.com/eslint/typescript-eslint-parser/commit/5421ceae59a10212562eedbafbd4bea116ca2b6e

File: src/language-js/needs-parens.js [hunk 3]
```diff
@@ -215,7 +215,11 @@ function needsParens(path, options) {
           if (!parent) {
             return false;
           }
-          if (parent.type === "ForStatement" && parent.init === node) {
+          if (
+            parent.type === "ForStatement" &&
+            // eslint-disable-next-line prettier-internal-rules/better-parent-property-check-in-needs-parens
```
Comment (thorn0) [inline src/language-js/needs-parens.js:220]: How about renaming the `parent` variable to `ancestor` to get rid of `eslint-disable`?

File: src/language-js/needs-parens.js (hunk 2 above)
Comment (fisker) [reply src/language-js/needs-parens.js:510]: Nice catch, I thought tests not covered.

File: src/language-js/needs-parens.js (hunk 3 above)
Comment (fisker) [reply src/language-js/needs-parens.js]: This function is not right

**Prettier 2.1.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBLAZgCnRCACVKPAIwEMAnASgG48B6OvAHgFoW8p8AHChAZwA6UITnJ5MkKHxh4A+ngC84zJUUA+PMCF4CWHPkIkKNeo1bse5fkIC+lFbVqVgNoSAA0ICFxipofZFAKcggAdwAFXilkEFIAG1DSAE8Az2JyUjAAazgYAGUeMEIAc2QYcgBXOE84AFtiOAATRqaAGVIoYorSYrgAMQhyWtIYX06Y0gqYCA8QAAsYWriAdTnUeD5CuDz+ddQAN3WkmLA+VLQpOHIYcIzi4eR0eL5qkAArPgAPACEM7Ny8qRanBWoQ4I9nq8Pp88iU4nAAIoVCDwCFxF6eSwvcgxMgNOKzLjkQgwZaoRowObIAAcAAZMSEXssMlwYkS4Nj9uDPABHZHwW7eAIoUh8FhQOBNJqzKx81BWW49B5IJ7o14vWqoMqVdVwxH88EqyGeGCkYhkilUpAAJhNGVQcRKAGEILVlSAOQBWWYVF4AFTNwtVGJA+yqAEkoC1YHkwMSfABBKN5GBJeFol42GxAA)
```sh
--parser babel
```

**Input:**
```jsx
if(foo in bar); // <-- no parens

for (const _ = (() => {
  if(foo in bar); // <-- parens
})(); ; ){}

```

**Output:**
```jsx
if (foo in bar); // <-- no parens

for (
  const _ = (() => {
    if ((foo in bar)); // <-- parens
  })();
  ;

) {}

```

Comment (fisker) [reply src/language-js/needs-parens.js]: https://github.com/prettier/prettier/issues/907#issuecomment-284304321

**Review by sosukesuzuki — APPROVED** [review]

**Pull request merged**